datafusion/optimizer/src/analyzer/type_coercion.rs

```diff
@@ -147,20 +148,24 @@ impl TreeNodeRewriter for TypeCoercionRewriter {
                 subquery,
                 negated,
             } => {
-                let expr_type = expr.get_type(&self.schema)?;
                 let new_plan = analyze_internal(&self.schema, &subquery.subquery)?;
+                let expr_type = expr.get_type(&self.schema)?;
                 let subquery_type = new_plan.schema().field(0).data_type();
-                let expr = if &expr_type == subquery_type {
-                    expr
-                } else {
-                    Box::new(expr.cast_to(subquery_type, &self.schema)?)
+                let result_type = comparison_coercion(&expr_type, subquery_type).ok_or(DataFusionError::Plan(
```

Comment:
maybe the common_type or coercion_type is better.
the result type of `in subquery` or `comparation` is `boolean`.